sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java

```diff
@@ -52,105 +57,142 @@ public HttpLoggingPolicy(HttpLogOptions httpLogOptions) {
+            || numericLogLevel > LogLevel.INFORMATIONAL.toNumeric()) {
+            return Mono.empty();
         }
 
-        if (httpLogLevel.shouldLogHeaders()) {
-            formatAllowableHeaders(httpLogOptions.getAllowedHeaderNames(), request.getHeaders(), logger);
+        StringBuilder requestLogMessage = new StringBuilder();
+        if (httpLogDetailLevel.shouldLogUrl()) {
+            requestLogMessage.append(String.format("--> %s %s%n", request.getHttpMethod(),
+                getRedactedUrl(request.getUrl())));
         }
 
-        //
-        Mono<Void> reqBodyLoggingMono = Mono.empty();
-        //
-        if (httpLogLevel.shouldLogBody()) {
+        addHeadersToLogMessage(request.getHeaders(), requestLogMessage, numericLogLevel);
+
+        Mono<String> requestLoggingMono = Mono.defer(() -> Mono.just(requestLogMessage.toString()));
+
+        if (httpLogDetailLevel.shouldLogBody()) {
             if (request.getBody() == null) {
-                logger.info("(empty body)");
-                logger.info("--> END {}", request.getHttpMethod());
+                requestLogMessage.append(String.format("(empty body)%n--> END %s%n", request.getHttpMethod()));
             } else {
-                boolean isHumanReadableContentType =
-                    !"application/octet-stream".equalsIgnoreCase(request.getHeaders().getValue("Content-Type"));
-                final long contentLength = getContentLength(logger, request.getHeaders());
-
-                if (contentLength < MAX_BODY_LOG_SIZE && isHumanReadableContentType) {
-                    try {
-                        Mono<byte[]> collectedBytes = FluxUtil.collectBytesInByteBufferStream(request.getBody());
-                        reqBodyLoggingMono = collectedBytes.flatMap(bytes -> {
-                            String bodyString = new String(bytes, StandardCharsets.UTF_8);
-                            bodyString = prettyPrintIfNeeded(
-                                logger,
-                                request.getHeaders().getValue("Content-Type"),
-                                bodyString);
-                            logger.info("{}-byte body:%n{}", contentLength, bodyString);
-                            logger.info("--> END {}", request.getHttpMethod());
-                            return Mono.empty();
-                        });
-                    } catch (Exception e) {
-                        reqBodyLoggingMono = Mono.error(e);
-                    }
+                String contentType = request.getHeaders().getValue("Content-Type");
+                long contentLength = getContentLength(logger, request.getHeaders());
+
+                if (bodyIsPrintable(contentType, contentLength)) {
+                    requestLoggingMono = FluxUtil.collectBytesInByteBufferStream(request.getBody()).flatMap(bytes ->
+                        Mono.just(requestLogMessage.append(String.format("%d-byte body:%n%s%n--> END %s%n",
```

Comment:
Avoid using `String.format()` as this code gets executed for every request (if logging is enabled) and the performance of `String.format()` is bad.

Also, the environment variable for logging might be set to a lower numeric value but the slf4j log level might be higher which means that all these strings will be created but not really logged.

Reply:
Removed all `String.format()`s, now using `StringBuilder.append` for all additions to the log message.